protocol/nats_jetstream/v3/protocol.go

```diff
@@ -0,0 +1,269 @@
+		return nil, err
+	}
+
+	if err := p.validateOptions(); err != nil {
+		return nil, err
+	}
+
+	var errConnection error
+	defer func() {
+		// close connection if an error occurs
+		if errConnection != nil {
+			p.Close(ctx)
+		}
+	}()
+
+	// if a URL was given create the nats connection
+	if p.conn == nil {
+		p.conn, errConnection = nats.Connect(p.url, p.natsOpts...)
+		if errConnection != nil {
+			return nil, errConnection
+		}
+	}
+
+	if p.jetStream, errConnection = jetstream.New(p.conn, p.jetSteamOpts...); errConnection != nil {
+		return nil, errConnection
+	}
+
+	return p, nil
+}
+
+// Send sends messages. Send implements Sender.Sender
+func (p *Protocol) Send(ctx context.Context, in binding.Message, transformers ...binding.Transformer) (err error) {
+	if p.sendSubject == "" {
+		return ErrNoSendSubject
+	}
+
+	defer func() {
+		if err2 := in.Finish(err); err2 != nil {
+			if err == nil {
+				err = err2
+			} else {
+				err = fmt.Errorf("failed to call in.Finish() when error already occurred: %s: %w", err2.Error(), err)
+			}
+		}
+	}()
+
+	if _, err = p.jetStream.StreamNameBySubject(ctx, p.sendSubject); err != nil {
+		return err
+	}
+
+	writer := new(bytes.Buffer)
+	header, err := WriteMsg(ctx, in, writer, transformers...)
+	if err != nil {
+		return err
+	}
+
+	natsMsg := &nats.Msg{
+		Subject: p.sendSubject,
+		Data:    writer.Bytes(),
+		Header:  header,
+	}
+
+	_, err = p.jetStream.PublishMsg(ctx, natsMsg, p.publishOpts...)
+
+	return err
+}
+
+// OpenInbound implements Opener.OpenInbound
+func (p *Protocol) OpenInbound(ctx context.Context) error {
+	p.subMtx.Lock()
+	defer p.subMtx.Unlock()
+
+	var consumeContext jetstream.ConsumeContext
+	var err error
+	if err = p.createJetstreamConsumer(ctx); err != nil {
+		return err
+	}
+	if consumeContext, err = p.jetstreamConsumer.Consume(p.MsgHandler, p.pullConsumeOpts...); err != nil {
+		return err
+	}
+
+	// Wait until external or internal context done
+	select {
+	case <-ctx.Done():
+	case <-p.internalClose:
+	}
+
+	// Finish to consume messages in the queue and close the subscription
+	if consumeContext != nil {
+		consumeContext.Drain()
+	}
+	return nil
+}
+
+// Receive implements Receiver.Receive.
+func (p *Protocol) Receive(ctx context.Context) (binding.Message, error) {
+	select {
+	case msgErr, ok := <-p.incoming:
+		if !ok {
+			return nil, io.EOF
+		}
+		return msgErr.msg, nil
+	case <-ctx.Done():
+		return nil, io.EOF
+	}
+}
+
+// MsgHandler implements nats.MsgHandler and publishes messages onto our internal incoming channel to be delivered
+// via r.Receive(ctx)
+func (p *Protocol) MsgHandler(msg jetstream.Msg) {
+	p.incoming <- msgErr{msg: NewMessage(msg)}
+}
+
+// Close implements Closer.Close
+func (p *Protocol) Close(ctx context.Context) error {
+	// Before closing, let's be sure OpenInbound completes
+	// We send a signal to close and then we lock on subMtx in order
+	// to wait OpenInbound to finish draining the queue
+	p.internalClose <- struct{}{}
```

Comment:
If OpenInbound is already done (ctx cancelled) this will deadlock

Reply:
The code does this:
`internalClose: make(chan struct{}, 1),`

I believe closing a channel will not block:
```
// The close built-in function closes a channel, which must be either
// bidirectional or send-only. It should be executed only by the sender,
// never the receiver, and has the effect of shutting down the channel after
// the last sent value is received. After the last value has been received
// from a closed channel c, any receive from c will succeed without
// blocking, returning the zero value for the channel element. The form
//
//	x, ok := <-c
//
// will also set ok to false for a closed and empty channel.
```

Reply:
Sorry, I somehow missed that you already had used a buffered channel here